dynamiqs/sparse.py

```diff
@@ -0,0 +1,202 @@
+            end = min(N, N + offset)
+            top = max(0, -offset)
+            bottom = top + end - start
+            if direction=='left':
+                out = out.at[top:bottom, :].add(
+                    diag[start:end, None] * matrix[start:end, :]
+                )
+            else:
+                out = out.at[:, start:end].add(
+                    jnp.transpose(diag[start:end, None] * jnp.transpose(matrix[:, top:bottom]))
+                )
+        return out
+
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _matmul_dia(self, matrix) -> Array:
+        N = matrix.diags.shape[1]
+
+        vector_list = []
+        offset_list = []
+
+        for offset, diag in zip(self.offsets, self.diags):
+            for matrix_offset, matrix_diag in zip(matrix.offsets, matrix.diags):
+
+                out_diag = jnp.zeros_like(diag)
+
+                sA, sB = max(0, -matrix_offset), max(0, matrix_offset)
+                eA, eB = min(N, N-matrix_offset), min(N, N+matrix_offset)
+
+                out_diag = out_diag.at[sB:eB].add(
+                    diag[sA:eA] * matrix_diag[sB:eB]
+                )
+
+                new_offset = offset + matrix_offset
+
+                vector_list.append(out_diag)
+                offset_list.append(new_offset)
+
+        out = jnp.vstack(vector_list)
+
+        return out, offset_list
+
+
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _add_dia(self, matrix):
+        self_offsets = list(self.offsets)
+        matrix_offsets = list(matrix.offsets)
+
+        offset_to_diag = {offset: diag for offset, diag in zip(self_offsets, self.diags)}
+
+        for offset, diag in zip(matrix_offsets, matrix.diags):
+            if offset in offset_to_diag:
+                offset_to_diag[offset] += diag
+            else:
+                offset_to_diag[offset] = diag
+
+        new_offsets = sorted(offset_to_diag.keys())
+        new_diagonals = jnp.array([offset_to_diag[offset] for offset in new_offsets])
+
+        return new_diagonals, new_offsets
+
+
+    def _dag(self):
+        """Returns the hermitian conjugate, call to_dense() to visualize."""
+        self.offsets = tuple(-1 * jnp.array(self.offsets))
+        self.diags = jnp.conjugate(self.diags)
+
+    @functools.partial(jax.jit, static_argnums=(0,))
+    def _add_dense(self, matrix: Array) -> Array:
+        """Sparse + Dense only using information about diagonals & offsets."""
+        N = matrix.shape[0]
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+
+            s = max(0, abs(offset))
+            e = min(N, N+abs(offset))
+
+            i = jnp.arange(e - s) if offset > 0 else jnp.arange(s, e)
+            j = jnp.arange(s, e) if offset > 0 else jnp.arange(e - s)
+
+            matrix = matrix.at[i, j].add(diag[start:end])
+
+        return matrix
+    
+
+    def _cleanup(self, diags, offsets):
```

Comment:
I think we should also add a test for this function, since it has non-straightforward logic.